### PR TITLE
Convert agent to Kotlin

### PR DESCRIPTION
**Pull request by DreierF**
- [ ] [No new Teamscale findings](https://build.cqse.eu/teamscale/delta.html#input/jacoco-client/?showMergeFindings=true)
- [ ] [Changes are tested adequately](https://build.cqse.eu/teamscale/tests.html#/jacoco-client/)
- [ ] USAGE.md updated in case of user-visible changes
- [x] CHANGELOG.md updated
- [x] build.gradle version number increased according to semantic versioning

### CODE REVIEW

File: javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt [hunk 1]
```diff
@@ -45,7 +43,7 @@ constructor(workingDirectory: Path) {
     @Throws(InstallationException::class)
     private fun validate() {
         try {
-            FileSystemUtils.mkdirs(backupPaths.backupDirectory.toFile())
+            backupPaths.backupDirectory.toFile().mkdirs()
         } catch (e: IOException) {
```
Comment (karottenreibe) [inline javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt]: mkdirs does not throw IO exceptions, it returns true or false. please handle accordingly

File: javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt [hunk 2]
```diff
@@ -73,7 +71,7 @@ constructor(workingDirectory: Path) {
         }
 
         try {
-            FileSystemUtils.mkdirs(wrapperPaths.defaultOutputDirectory.toFile())
+            wrapperPaths.defaultOutputDirectory.toFile().mkdirs()
```
Comment (karottenreibe) [inline javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt]: same here

File: javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt [hunk 3]
```diff
@@ -92,7 +90,7 @@ constructor(workingDirectory: Path) {
         }
 
         try {
-            FileSystemUtils.writeFileUTF8(backupPaths.ftypeMapping.toFile(), readCurrentJnlpFtype())
+            backupPaths.ftypeMapping.toFile().writeText(readCurrentJnlpFtype())
```
Comment (karottenreibe) [inline javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt]: does writetext actually throw IO Exceptions? the documentation does not say. please check and adjust catch clause if necessary

File: javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt [hunk 4]
```diff
@@ -138,7 +136,7 @@ constructor(workingDirectory: Path) {
             FileOutputStream(wrapperPaths.configProperties.toFile()).use { outputStream ->
                 properties.store(
```
Comment (karottenreibe) [inline javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt]: put on one line?

File: report-generator/src/main/kotlin/com/teamscale/report/util/AntPatternUtils.kt [hunk 5]
```diff
@@ -1,16 +1,170 @@
...
+ * We implement a special version where a trailing '.' can be used to only match
+ * files without file extension (i.e. file names without dot).
  */
 object AntPatternUtils {
 
     /** Converts an ANT pattern to a regex pattern.  */
     @Throws(PatternSyntaxException::class)
     fun convertPattern(antPattern: String, caseSensitive: Boolean): Pattern {
-        return org.conqat.lib.commons.filesystem.AntPatternUtils.convertPattern(antPattern, caseSensitive)
+        var normalizedAntPattern = normalizePattern(antPattern)
```
Comment (karottenreibe) [inline report-generator/src/main/kotlin/com/teamscale/report/util/AntPatternUtils.kt]: Why can't we reuse the one from the commons anymore? is that jar not java 7 compatible?

Comment (karottenreibe) [reply report-generator/src/main/kotlin/com/teamscale/report/util/AntPatternUtils.kt]: if we indeed can't reuse, please also copy the tests here

File: report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+object FileSystemUtils {
+
+    /** Unix file path separator  */
+    val UNIX_SEPARATOR = '/'
+
+    /**
+     * Replace platform dependent separator char with forward slashes to create
+     * system-independent paths.
+     */
+    fun normalizeSeparators(path: String): String {
```
Comment (karottenreibe) [inline report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt]: please reuse this. I've seen at least one other place where this code is cloned

File: report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package com.teamscale.report.util
+
+import java.io.File
+import java.io.IOException
+
+object FileSystemUtils {
+
+    /** Unix file path separator  */
+    val UNIX_SEPARATOR = '/'
```
Comment (karottenreibe) [inline report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt]: i've seen at least one other place where this char literal is cloned. plesae find and reuse

File: agent/src/main/java/com/teamscale/jacoco/agent/AgentOptionsParser.kt [hunk 8]
```diff
@@ -105,7 +105,7 @@ class AgentOptionsParser(
             }
             "interval" -> {
                 try {
-                    options.dumpIntervalInMinutes = Integer.parseInt(value)
+                    options.dumpIntervalInMinutes = java.lang.Long.parseLong(value)
```
Comment (karottenreibe) [inline agent/src/main/java/com/teamscale/jacoco/agent/AgentOptionsParser.kt]: use value.toLong() instead?

File: agent/src/main/java/com/teamscale/jacoco/agent/Main.kt [hunk 9]
```diff
@@ -28,7 +28,7 @@ class Main {
      * exception if the arguments are not valid. Then runs the specified command.
      */
     @Throws(Exception::class)
-    private fun parseCommandLineAndRun(args: Array<String>) {
+    internal fun parseCommandLineAndRun(args: Array<String>) {
```
Comment (karottenreibe) [inline agent/src/main/java/com/teamscale/jacoco/agent/Main.kt]: why the change to internal?

File: agent/src/main/java/com/teamscale/jacoco/agent/Main.kt [hunk 10]
```diff
@@ -63,19 +63,16 @@ class Main {
...
-        val help: Boolean = false
+        var help: Boolean = false
 
     }
 
     companion object {
 
         /** Version of this program.  */
-        private val VERSION: String
+        private val bundle = ResourceBundle.getBundle("com.teamscale.jacoco.agent.app")!!
```
Comment (karottenreibe) [inline agent/src/main/java/com/teamscale/jacoco/agent/Main.kt]: can be inlined. also the comment is wrong

**Comment by karottenreibe** [comment]
Big change! So naturally there are some comments :-) Please also:

- fix merge conflicts
- comment on how you tested these changes (we should at least run the agent once with a JRE 7 to make sure it still produces traces etc. The more code paths this test touches the better)
- can we add an automated test to ensure the agent works under Java 7? Or can we configure IntelliJ to flag any Java 8+ API usage as an error? I fear it's really easy otherwise to use Duration, Predicate or one of the other Java 8+ APIs by accident in the future
- update documentation to say we support Java 7

File: report-generator/src/main/java/com/teamscale/report/testwise/model/builder/FileCoverageBuilder.kt [hunk 11]
```diff
@@ -83,7 +82,7 @@ class FileCoverageBuilder
             }
 
             val linesList = ArrayList(lines)
-            linesList.sort(Comparator { obj, anotherInteger -> obj.compareTo(anotherInteger) })
+            linesList.sortWith(Comparator { obj, anotherInteger -> obj.compareTo(anotherInteger) })
```
[the remark this answers is not in the document]
Comment (DreierF) [reply report-generator/src/main/java/com/teamscale/report/testwise/model/builder/FileCoverageBuilder.kt]: I think this is a leftover from the LineRange to Int conversion. Repalced with:
```
val linesList = lines.sorted().toMutableList()
```

File: agent/src/main/java/com/teamscale/jacoco/agent/AgentOptionsParser.kt [hunk 12]
```diff
@@ -379,17 +364,17 @@ class AgentOptionsParser(
...
         val workingDir = workingDirectory.absoluteFile
-        val basePath = workingDir.toPath().resolve(baseDir).normalize().toAbsolutePath()
+        val basePath = workingDir.resolve(baseDir).normalize().absoluteFile
 
         val pathMatcher = AntPatternUtils.convertPattern(pattern, false)
-        val filter = { path ->
+        val filter = { path: File ->
             pathMatcher
-                .matcher(FileSystemUtils.normalizeSeparators(basePath.relativize(path).toString())).matches()
+                .matcher(FileSystemUtils.normalizeSeparators(basePath.relativeTo(path).toString())).matches()
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/AgentOptionsParser.kt]: No, the Paths API is exactly specified the other way round. We also have test for this :)

File: agent/src/main/java/com/teamscale/jacoco/agent/DelayedLogger.kt [hunk 13]
```diff
@@ -12,43 +11,35 @@ import java.util.ArrayList
...
-        logActions.forEach { action -> action.log(logger) }
-    }
-
-    /** An action to be executed on a logger.  */
-    private interface ILoggerAction {
-
-        /** Executes the action on the given logger.  */
-        fun log(logger: Logger)
-
+        logActions.forEach { action -> action.invoke(logger) }
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/DelayedLogger.kt]: Completely forgot about that syntax

File: agent/src/main/java/com/teamscale/jacoco/agent/Main.kt [hunk 14]
```diff
@@ -87,7 +87,7 @@ class Main {
         }
 
         /** Shows an informative error and help message. Then exits the program.  */
-        private fun handleInvalidCommandLine(jCommander: JCommander, message: String) {
+        private fun handleInvalidCommandLine(jCommander: JCommander, message: String?) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/Main.kt]: The reason was that we call it e.g. here: `handleInvalidCommandLine(jCommander, e.message)` where Exception#message is nullable.
But your objection makes sense.
Replaced the call with `handleInvalidCommandLine(jCommander, "${e.message}")` and made message non-null

File: agent/src/main/java/com/teamscale/jacoco/agent/commandline/Validator.kt [hunk 15]
```diff
@@ -16,7 +16,7 @@ import java.util.ArrayList
 class Validator {
 
     /** The found validation problems in the form of error messages for the user.  */
-    private val messages = ArrayList<String>()
+    private val messages = ArrayList<String?>()
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/commandline/Validator.kt]: Same here

File: agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt [hunk 16]
```diff
@@ -11,23 +11,24 @@ import java.io.IOException
 class UploadStoreException : Exception {
 
     /** Constructor  */
-    constructor(message: String, e: Exception) : super(message, e) {}
+    constructor(message: String, e: Exception) : super(message, e)
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt]: I thought I must not convert the code to idiomatic Kotlin 😜

[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt]: I don't see any way of rewriting this with primary constructor and keeping the current behavior.
The problem is the optional e argument. Currently the exception cause (e) is implicitly set to this if not specified (not null as I naively thought), but Kotlin does obviously not allow me to use this as default value for e.
Am I missing something here?

File: agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt [hunk 17]
```diff
@@ -11,23 +11,24 @@ import java.io.IOException
...
+
+    companion object {
+        private fun createResponseMessage(message: String, response: Response<ResponseBody>): String {
+            return try {
+                val errorBodyMessage = response.errorBody()!!.string()
+                String.format("%s (%s): \n%s", message, response.code(), errorBodyMessage)
+            } catch (e: IOException) {
+                message
+            } catch (e: NullPointerException) {
+                message
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt]: Can be handled better in Kotlin of course

File: build.gradle [hunk 18]
```diff
@@ -38,12 +38,12 @@ subprojects {
...
-	sourceCompatibility = 1.8
-	targetCompatibility = 1.8
+	sourceCompatibility = 1.7
+	targetCompatibility = 1.7
 
-	tasks.withType(JavaCompile) {
-		sourceCompatibility = "1.8"
-		targetCompatibility = "1.8"
+	tasks.named('compileJava').configure {
+		sourceCompatibility = "1.7"
```
[the remark this answers is not in the document]
Comment (DreierF) [reply build.gradle]: only 1.6 and 1.8 is allowed here. 1.6 is technically the same as 1.6 (on bytecode level)

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/closure/ClosureTestwiseCoverageGenerator.kt [hunk 19]
```diff
@@ -78,18 +75,17 @@ class ClosureTestwiseCoverageGenerator
...
             }
 
-            val coveredFile = File(fileNameAndExecutedLines.first)
-            val coveredLines = fileNameAndExecutedLines.second
+            val coveredFile = File(fileName)
             val path = Optional.ofNullable(coveredFile.parent).orElse("")
             val fileCoverage = FileCoverageBuilder(path, coveredFile.name)
             for (i in coveredLines.indices) {
-                if (coveredLines[i] != null && coveredLines[i]) {
+                if (coveredLines[i] == true) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply report-generator/src/main/kotlin/com/teamscale/report/testwise/closure/ClosureTestwiseCoverageGenerator.kt]: This stems directly from the Google Closure Coverage files, which specify for each line if it was covered = `true` or `null` if not covered.
```
"executedLines": [
    [
      null,
      true,
      true,...
```

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/model/builder/TestCoverageBuilder.kt [hunk 20]
```diff
@@ -22,18 +22,15 @@ class TestCoverageBuilder
...
     /** Returns true if there is no coverage for the test yet.  */
     val isEmpty: Boolean
         get() = pathCoverageList.isEmpty()
 
     /** Adds the [FileCoverageBuilder] to into the map, but filters out file coverage that is null or empty.  */
     fun add(fileCoverage: FileCoverageBuilder?) {
-        if (fileCoverage == null || fileCoverage.isEmpty
-            || fileCoverage.fileName == null || fileCoverage.path == null
-        ) {
+        if (fileCoverage == null || fileCoverage.isEmpty) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply report-generator/src/main/kotlin/com/teamscale/report/testwise/model/builder/TestCoverageBuilder.kt]: Yes, the fileName and path are non-nullable now, so we don't need this sanity check anymore

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/model/builder/TestCoverageBuilder.kt [hunk 21]
```diff
@@ -22,18 +22,15 @@ class TestCoverageBuilder
...
             return
         }
         val pathCoverage = pathCoverageList.computeIfAbsent(fileCoverage.path) { PathCoverageBuilder(it) }
```
[the remark this answers is not in the document]
Comment (DreierF) [reply report-generator/src/main/kotlin/com/teamscale/report/testwise/model/builder/TestCoverageBuilder.kt]: path is non-nullable

File: javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt [hunk 22]
```diff
@@ -225,9 +224,9 @@ constructor(workingDirectory: Path) {
         }
     }
 
-    private inner class BackupPaths(private val backupDirectory: Path) {
-        private val securityPolicy: Path
-        private val ftypeMapping: Path
+    private inner class BackupPaths(internal val backupDirectory: Path) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt]: Because private fields of inner classes are accessible to the surrounding class in Java, but are hidden from it in Kotlin. So I needed to make them internal.

File: agent/src/main/java/com/teamscale/jacoco/agent/commandline/Validator.kt [hunk 23]
```diff
@@ -42,15 +40,15 @@ class Validator {
      * message.
      */
     fun isTrue(condition: Boolean, message: String) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/commandline/Validator.kt]: I woudn't. I think it's more readable as is

File: agent/src/main/java/com/teamscale/jacoco/agent/commandline/Validator.kt [hunk 24]
```diff
@@ -42,15 +40,15 @@ class Validator {
...
-        ensure { CCSMAssert.isTrue(condition, message) }
+        ensure { require(condition) { message } }
     }
 
     /**
      * Checks that the given condition is `false` or adds the given error
      * message.
      */
     fun isFalse(condition: Boolean, message: String) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/commandline/Validator.kt]: Same here

File: javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt (hunk 3 above)
Comment (DreierF) [reply javaws-wrapper/src/main/java/com/teamscale/jacoco/javaws/WindowsInstallation.kt]: It calls `FileOutputStream` internally, which may throw a `FileNotFoundException` and `write` which throws an `IOException`.
https://discuss.kotlinlang.org/t/in-future-could-kotlin-have-checked-exception/1579/15

File: report-generator/src/main/kotlin/com/teamscale/report/util/AntPatternUtils.kt (hunk 5 above)
Comment (DreierF) [reply report-generator/src/main/kotlin/com/teamscale/report/util/AntPatternUtils.kt]: That's exactly the case. I needed to replace our lib commons as it is compiled against Java 8.

File: report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt (hunk 6 above)
Comment (DreierF) [reply report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt]: The same logic is used in javaws, but I didn't wanted to add the whole report-generator + jacoco-runtime+teamscale-client dependency tree just to share this single line of code.

File: report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt (hunk 7 above)
Comment (DreierF) [reply report-generator/src/main/kotlin/com/teamscale/report/util/FileSystemUtils.kt]: It's used in a few places, but with different semantics (except the javaws occurrence).
It's used as test path separator and as separator for package name elements in other places.

File: agent/src/main/java/com/teamscale/jacoco/agent/Main.kt (hunk 9 above)
Comment (DreierF) [reply agent/src/main/java/com/teamscale/jacoco/agent/Main.kt]: Cannot remember. Set to private

**Comment by DreierF** [comment]
I ran the agent together with the sample project on Windows running a Java 7 JDK as this was the only platform I could get a JDK7 to run.
I enabled an inspection in IntelliJ that checks for API usages that are marked with `@since(1.8)` or higher.
I think we cannot easily switch the jdk version on Travis as the test code still runs against Java 8 and the impacted-test-executor still needs Java 8. This is kind of very unfortunate situation.
I plan to migrate the remaining impacted-test-executor to Kotlin as well, but Albert is currently working on some major changes in there and I'd like to delay the conversion to avoid merge conflicts.
I'm open for any further ideas on how to improve the situation.
RFR

**Comment by DreierF** [comment]
Merged master in and resolved merge conflicts

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingClassAnalyzer.java [hunk 25]
```diff
@@ -39,7 +37,17 @@ public void visitSource(String source, String debug) {
 	@Override
 	public MethodProbesVisitor visitMethod(final int access, final String name,
 										   final String desc, final String signature, final String[] exceptions) {
-		return new CachingMethodAnalyzer(classCoverageLookup, Filters.ALL, this);
+		final CachingInstructionsBuilder builder = new CachingInstructionsBuilder(classCoverageLookup);
+
+		return new MethodAnalyzer(builder) {
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingClassAnalyzer.java]: please comment why overriding that one method is necessary. This might be confusing for the "casual" reader ;-)

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/jacoco/cache/ProbesCache.kt [hunk 26]
```diff
@@ -72,9 +70,9 @@ class ProbesCache
             )
             return null
         }
-        return if (!executionData.hasHits()) {
-            null
-        } else classCoverageLookups[classId]?.getFileCoverage(executionData, logger)
-
+        return when {
+            !executionData.hasHits() -> null
```
Comment (karottenreibe) [inline report-generator/src/main/kotlin/com/teamscale/report/testwise/jacoco/cache/ProbesCache.kt]: please invert condition for readability

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/model/builder/TestInfoBuilder.kt [hunk 27]
```diff
@@ -47,14 +47,15 @@ class TestInfoBuilder
...
-        val testInfo = if (execution != null) {
-            TestInfo(
-                uniformPath, details!!.sourcePath, details!!.content,
-                execution!!.durationSeconds,
-                execution!!.result, execution!!.message
-            )
-        } else {
-            TestInfo(uniformPath, details!!.sourcePath, details!!.content, null, null, null)
+        val testInfo = when {
+            execution != null -> {
```
Comment (karottenreibe) [inline report-generator/src/main/kotlin/com/teamscale/report/testwise/model/builder/TestInfoBuilder.kt]: please invert condition for readability

File: agent/src/main/java/com/teamscale/jacoco/agent/Main.kt [hunk 28]
```diff
@@ -65,7 +65,7 @@ class Main {
 
         /** Shows the help message.  */
         @Parameter(names = arrayOf("--help"), help = true, description = "Shows all available command line arguments.")
-        private val help: Boolean = false
+        val help: Boolean = false
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/src/main/java/com/teamscale/jacoco/agent/Main.kt]: unaddressed comment

File: agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/src/main/java/com/teamscale/jacoco/agent/store/UploadStoreException.kt]: I didn't know this. I always pass null. does that simply not work (exception thrown or something)?

File: report-generator/src/main/kotlin/com/teamscale/report/testwise/closure/ClosureTestwiseCoverageGenerator.kt (hunk 19 above)
[the remark this answers is not in the document]
Comment (karottenreibe) [reply report-generator/src/main/kotlin/com/teamscale/report/testwise/closure/ClosureTestwiseCoverageGenerator.kt]: ok, then please add that to the class comment so it's clear :-)

**Comment by karottenreibe** [comment]
please resolve conflicts as well

**Comment by DreierF** [comment]
RFR

**Comment by karottenreibe** [comment]
Closing this for now as noone's working on this and it has loads of merge conflicts. Reopen if needed
